### ISSUE TITLE
Connection to ipv6 host is successful but results in 'Websocket connection offline' errors.

### ISSUE THREAD

**Issue by atcronin** [MediaBrowser/Emby.Kodi#471]
this line in websocket.py is the issue:
```python
124       self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
raises the Exception on self.sock.connect
```python
[Errno 11001] getaddrinfo failed
```
need to use AF_INET6 for ipv6 addresses.

changing it to:
```python
address_family = socket.getaddrinfo(hostname, None)[0][0]
self.sock = socket.socket(address_family, socket.SOCK_STREAM)
```
will determine the address family from the first resolved address.
_getaddrinfo returns a list of (family, socktype, proto, canonname, sockaddr)_

The lines will need to come after 

```python
136        hostname = parsed.hostname
```

perhaps:

```python
def establish_connection(self):
    self.close(False)

    url = f"{self.EmbyServerSocketUrl}/embywebsocket?api_key={self.EmbyServer.ServerData['AccessToken']}&device_id={utils.device_id}"

    # Parse URL
    scheme, url = url.split(":", 1)
    parsed = urlparse(url, scheme="http")
    resource = parsed.path

    if parsed.query:
        resource += f"?{parsed.query}"

    hostname = parsed.hostname
    port = parsed.port
    is_secure = scheme == "wss"

    if not port:
        if scheme == "http":
            port = 80
        else:
            port = 443

    # Connect
    try:
        address_family = socket.getaddrinfo(hostname, None)[0][0]
        self.sock = socket.socket(address_family, socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        self.sock.connect((hostname, port))
```

tested on Kodi 19.4 + 7.13.7
with ipv4 notation and dns, ipv6 dns.

_another issue arises when using ipv6 notation. 
the plugin connects when wrapping the serverurl address in square brackets per rfc2732, but while establishing the websocket the address hostname is used in the header_str without the square brackets resulting in a "400 Bad Request"
creating a temporary hostname in brackets for header_str fixes that issue as far as i can tell._
```python
hostname_esc = hostname
if ':' in hostname:
    hostname_esc = "[%s]" % hostname
header_str = "GET %s HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nHost: %s:%d\r\nSec-WebSocket-Key: %s\r\nSec-WebSocket-Version: 13\r\n\r\n" % (resource, hostname_esc, port, EncodingKey)
```

**Comment by atcronin** [comment]
this function from urlib3/util/connection.py might provide an example of a more robust method for establishing the socket connection, or just replace the existing method entirely.
just in case the scenario of a domain name returning multiple addresses but for whatever reason the first address is unreachable but a later one is. 

```Python
# This function is copied from socket.py in the Python 2.7 standard
# library test suite. Added to its signature is only `socket_options`.
# One additional modification is that we avoid binding to IPv6 servers
# discovered in DNS if the system doesn't have IPv6 functionality.
def create_connection(
    address,
    timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
    source_address=None,
    socket_options=None,
):
    """Connect to *address* and return the socket object.

    Convenience function.  Connect to *address* (a 2-tuple ``(host,
    port)``) and return the socket object.  Passing the optional
    *timeout* parameter will set the timeout on the socket instance
    before attempting to connect.  If no *timeout* is supplied, the
    global default timeout setting returned by :func:`socket.getdefaulttimeout`
    is used.  If *source_address* is set it must be a tuple of (host, port)
    for the socket to bind as a source address before making the connection.
    An host of '' or port 0 tells the OS to use the default.
    """

    host, port = address
    if host.startswith("["):
        host = host.strip("[]")
    err = None

    # Using the value from allowed_gai_family() in the context of getaddrinfo lets
    # us select whether to work with IPv4 DNS records, IPv6 records, or both.
    # The original create_connection function always returns all records.
    family = allowed_gai_family()

    try:
        host.encode("idna")
    except UnicodeError:
        return six.raise_from(
            LocationParseError(u"'%s', label empty or too long" % host), None
        )

    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
        af, socktype, proto, canonname, sa = res
        sock = None
        try:
            sock = socket.socket(af, socktype, proto)

            # If provided, set socket level options before connecting.
            _set_socket_options(sock, socket_options)

            if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
                sock.settimeout(timeout)
            if source_address:
                sock.bind(source_address)
            sock.connect(sa)
            return sock

        except socket.error as e:
            err = e
            if sock is not None:
                sock.close()
                sock = None

    if err is not None:
        raise err

    raise socket.error("getaddrinfo returns an empty list")
```

**Comment by quickmic** [comment]
Thanks, I'll review the issue.

**Comment by quickmic** [comment]
I'll add a fix in next version.

**Comment by quickmic** [comment]
Fixed in 8.2.12
